### ISSUE TITLE
Text output + fix ("count" is protected?)

### ISSUE THREAD

**Issue by qubodup** [berwynhoyt/wordcount-sections#1]
Thank you so much! I couldn't find anything for Word or Writer but I was able to modify this a little to output a CSV string into the file itself (at the end) so I can then copy it to a txt file, open it in Calc and copy from there to the sheet I need it for.

Ideally this would create a calc file and populate fields but I won't be able to invest the time I'm afraid.

"count" was used as a variable but that caused issues, so I renamed it.

So here is the code I ended up with:

```
REM  *****  BASIC  *****

' Count words between bookmarked sections -- attached to hotkey Ctrl-Alt-W
' Create insertable wordcount fields for the wordcounts under Insert -> Fields -> More Fields... -> Variables -> User Field
' If there are bookmarks, update the field counts every time the document is saved

' Developed using examples at: https://wiki.documentfoundation.org/Documentation/BASIC_Guide#Searching_for_Text_Portions
' And BASIC API reference at: https://api.libreoffice.org/docs/idl/ref/index.html

Sub Main
	REM recount_on_save() 'Setup save action to force re-count prior to save
	result = bookmark_wordcounter()

	REM prepare text writing stuff because I dont know how to create a calc sheet and dont want to deal with freaking cell navigation, just paste as cvs or something
	Dim oDoc As Object
	oDoc = ThisComponent
	Dim oText As Object
	oText = oDoc.getText()
	oText.insertString( oText.End, result, False)
	
End Sub

Function bookmark_wordcounter()
	doc = ThisComponent
	bookmarks = doc.Bookmarks
	n_bookmarks = bookmarks.getCount()

	cursor = doc.Text.createTextCursor
	cursor.gotoStart(false)

	' Count words at beginning of document up to the first bookmark to include in total
	results = "WORDCOUNT:" & Chr(10)
	if n_bookmarks > 0 then
		bookmark = bookmarks.getByIndex(0)
		cursor.gotoRange(bookmark.Anchor, true) ' true = expand selection
	else
		cursor.gotoEnd(true)
	end if
	counter = wordcount_string(cursor.String)
	total = counter
	'Uncomment if you want to show # words before the first bookmark
	'results = results & "Beginning" & "=" & count & ", "

	' Count words after each bookmark
	for i=1 to n_bookmarks
		name = bookmark.name
		cursor.gotoRange(bookmark.Anchor, False) ' true = expand selection
		if i < n_bookmarks then
			bookmark = bookmarks.getByIndex(i)
			cursor.gotoRange(bookmark.Anchor, true) ' true = expand selection
		else
			cursor.gotoEnd(true)
		end if
		counter = wordcount_string(cursor.String)
		total = total + counter
		update_field("Wordcount_" & name, counter)
		results = results & name & "," & counter & Chr(10)
	next i

	' Count words in footnotes so total matches libreoffice total
	fnotes=doc.getFootNotes()
	if fnotes.hasElements() then
		fnotecount=0
		for nfnotes=0 to fnotes.getCount()-1
			thisnote=fnotes.getbyIndex(nfnotes)
			fnotecount=fnotecount+wordcount_string(thisnote.getString())
		next nfnotes
		total = total + fnotecount
		update_field("Wordcount_Footnotes", fnotecount)
		results = results & "Footnotes," & fnotecount & Chr(10)
	end if

	results = results & "WordcountTotal" & "," & total & Chr(10)
	update_field("Wordcount_Total", total)

	' Warn if my calculation is different from libreoffice's
	' but allow some play in case the user puts a bookmark in the middle of a word which will effectively count it twice (once in each section)
	libreoffice_total = doc.WordCount
	if libreoffice_total > total or libreoffice_total + n_bookmarks < total then
		results = results & "LibreofficeTotal," & libreoffice_total & vbcrlf
	end if

	bookmark_wordcounter = results
	
End Function

Function wordcount_string(str)
	' Count words using Libreoffice's WORD_COUNT break iterator
	dim locale as new com.sun.star.lang.Locale
	
	' Treat n-dash and m-dash as spaces.
	' These two characters are the default in Tools -> Options -> Writer -> General -> Wordcount -> Additional Separators
	' If I knew how to fetch that option data, I'd use it here instead of hard-coding it.
	str = replace(str, "–", " ")
	str = replace(str, "—", " ")

	brk=createUnoService("com.sun.star.i18n.BreakIterator")
	locale.Language="en"
	boundary_type = com.sun.star.i18n.WordType.WORD_COUNT

	numwords=0
	startpos=0
	nextwd=brk.getWordBoundary(str, startpos, locale, boundary_type, TRUE)
	do while nextwd.endPos > nextwd.startPos
		'Include the following line to display each word as it is counted -- for debug purposes
		'msgbox(nextwd.startPos & " " & nextwd.endPos & ": " & mid(str, nextwd.startPos+1, nextwd.endPos-nextwd.startPos))
		numwords=numwords+1
		nextwd=brk.nextWord(str, nextwd.startPos, locale, boundary_type)
	loop
	wordcount_string=numwords
End Function

Function update_field(varname, value)
	' Update user field variable `varname`, creating it if necessary
	doc = ThisComponent
	user_path = "com.sun.star.text.fieldmaster.User"
	var_path = user_path & "." & varname
	on error goto ErrorHandler
		var = doc.getTextFieldMasters.getByName(var_path)
	on error goto 0
	var.Value = value
	exit function
ErrorHandler:
	var = doc.createInstance(user_path)
	var.Name = varname
	resume next
End Function

Sub on_save( oEvent As Object )
	' This subroutine gets connected to the "Save Document" event of LibreOffice, OnSaveAs.
	if ThisComponent.Bookmarks.getCount() > 0 then
		bookmark_wordcounter()
	end if
End Sub

Sub recount_on_save()
	' Call this macro to connect the LibreOffice "Save Document" event to the Basic macro `on_save()`.
	Dim aProps(1) As New com.sun.star.beans.PropertyValue
	aProps(0).Name		= "EventType"
	aProps(0).Value		= "Script"
	aProps(1).Name		= "Script"
	aProps(1).Value		= "vnd.sun.star.script:Wordcounter.wordcount_sections.bookmark_wordcounterExport?language=Basic&location=application"

	Dim oGlobalEventBroadcaster As Object
	oGlobalEventBroadcaster = GetDefaultContext().getByName( "/singletons/com.sun.star.frame.theGlobalEventBroadcaster" )
	oGlobalEventBroadcaster.Events.replaceByName("OnSave", aProps())
End Sub

```

I tried stripping some stuff (I didn't need any of the on_save things) but when I did, the LibreOffice wordcount came out nonsensical, so I just undid my attempts at "cleanup".

Feel free to treat any changes as public domain or cc0 and integrate them without credits.

PS: note to self, for when I need this in one year and will have forgotten all about it - ctrl-alt-w actually works! But you can also just go to the code editor, overwrite all the code with the above, put the cursor in Main() and press F5.

**Comment by berwynhoyt** [comment]
I'm glad you found it to be useful.

I'm interested to hear that you had trouble with the variable being named 'count'. I didn't have this issue. Can you tell me anything more about it? Did the original version of the addon work without any changes? You mention both Word and Writer. Are you running this in Microsoft Word or Libreoffice Writer?

Thanks,
Berwyn

**Comment by qubodup** [comment]
Hey, thank you for responding. I can't reproduce the issue with count any more. I think I downloaded the extension before the update on Jan 15th https://github.com/berwynhoyt/wordcount-sections/commit/ab4be1452da64be7368e31baacb15cd9d65d12ef and only installed and used it a month later. "My" code has a 'Main' function - that helped me potentially figure this one out today.

I had a problem with calc throwing an error whenever I saved a file, opening the macro editor and jumping to some line with doc specific code I think. Can't reproduce that one correctly either any more.

What I wanted was a spreadsheet of the word count of each section (headers). This script worked on the docx I used because that one had sections generated by TOC/table of contents.
